BlacklightCatalog.lua

```diff
@@ -402,16 +415,43 @@ function GridFocusedRowChanged(sender, args)
 	end
 end
 
+function GridRowDoubleClicked(sender, args)
+	local gridHitInfo = itemsGridView:CalcHitInfo(itemsGridControl:PointToClient(Types["System.Windows.Forms.Cursor"].Position));
+
+	if gridHitInfo.HitTest == Types["DevExpress.XtraGrid.Views.Grid.ViewInfo.GridHitTest"].RowCell then
+		DoItemImport();
+	end
+end
+
 function RetrieveItems()
 	log:Debug("Retrieving items");
 
-	local xmlRecord = DownloadString(tostring(CatalogForm.Browser.Address) .. ".xml");
+	local xmlRecordUrl = CatalogForm.Browser.Address .. ".xml";
+
+	-- Cache XML catalog record by URL if it hasn't been already.
+	if not xmlRecordsCache[xmlRecordUrl] then
+        xmlRecordsCache[xmlRecordUrl] = DownloadString(xmlRecordUrl);
+    end
+
+	local xmlRecord = Types["System.Xml.XmlDocument"]();
+    xmlRecord:LoadXml(xmlRecordsCache[xmlRecordUrl]);
 
 	CatalogForm.BibGrid.GridControl.MainView:BeginDataUpdate();
 	CatalogForm.ItemsGrid.GridControl.MainView:BeginDataUpdate();
 
-	CatalogForm.BibGrid.GridControl.DataSource = BuildBibDataSource(xmlRecord);
-	CatalogForm.ItemsGrid.GridControl.DataSource = BuildItemsDataSource(xmlRecord);
+	local bibSuccess, bibErr = pcall(function()
+		CatalogForm.BibGrid.GridControl.DataSource = BuildBibDataSource(xmlRecord);
+	end);
+	if not bibSuccess then
+		log:Error("Problem encountered while building the bib datasource. Error details:\n" .. TraverseError(bibErr));
+	end
+
+	local itemsSuccess, itemsErr = pcall(function()
```

Comment:
Should the item record only load if the bib record was loaded? It doesn't seem to short-circuit out when bibSuccess if false.